commit 8641a50a832f6f0461af5accea0edfbc9ce985a1
Author: bertin <macportsuser@macports.org>
Date:   Thu May 22 16:10:07 2025 +0200

    /opt/local/site-ports/sysutils/MacPorts/files/patch-record-patches-in-statefile.diff

diff --git a/src/port1.0/portpatch.tcl b/src/port1.0/portpatch.tcl
index 42dbbd227..3e61a230f 100644
--- a/src/port1.0/portpatch.tcl
+++ b/src/port1.0/portpatch.tcl
@@ -68,7 +68,7 @@ proc portpatch::build_getpatchtype {args} {
 }
 
 proc portpatch::patch_main {args} {
-    global UI_PREFIX
+    global UI_PREFIX target_state_fd
 
     # First make sure that patchfiles exists and isn't stubbed out.
     if {![exists patchfiles] || [option patchfiles] eq ""} {
@@ -96,18 +96,24 @@ proc portpatch::patch_main {args} {
     catch {set xzcat "[findBinary xz $portutil::autoconf::xz_path] -dc"}
 
     foreach patch $patchlist {
-        ui_info "$UI_PREFIX [format [msgcat::mc "Applying %s"] [file tail $patch]]"
-        switch -- [file extension $patch] {
-            .Z -
-            .gz {command_exec patch "$gzcat \"$patch\" | (" ")"}
-            .bz2 {command_exec patch "$bzcat \"$patch\" | (" ")"}
-            .xz {
-                if {[info exists xzcat]} {
-                    command_exec patch "$xzcat \"$patch\" | (" ")"
-                } else {
-                    return -code error [msgcat::mc "xz binary not found; port needs to add 'depends_patch bin:xz:xz'"]
-                }}
-            default {command_exec patch "" "< '$patch'"}
+        set pfile [file tail $patch]
+        if {![check_statefile patch $pfile $target_state_fd]} {
+            ui_info "$UI_PREFIX [format [msgcat::mc "Applying %s"] [file tail $patch]]"
+            switch -- [file extension $patch] {
+                .Z -
+                .gz {command_exec patch "$gzcat \"$patch\" | (" ")"}
+                .bz2 {command_exec patch "$bzcat \"$patch\" | (" ")"}
+                .xz {
+                    if {[info exists xzcat]} {
+                        command_exec patch "$xzcat \"$patch\" | (" ")"
+                    } else {
+                        return -code error [msgcat::mc "xz binary not found; port needs to add 'depends_patch bin:xz:xz'"]
+                    }}
+                default {command_exec patch "" "< '$patch'"}
+            }
+            write_statefile patch $pfile $target_state_fd
+        } else {
+            ui_info "$UI_PREFIX [format [msgcat::mc "Skipping already applied %s"] $pfile]"
         }
     }
     return 0
diff --git a/src/port1.0/portutil.tcl b/src/port1.0/portutil.tcl
index 14e9b9f4a..7192d4592 100644
--- a/src/port1.0/portutil.tcl
+++ b/src/port1.0/portutil.tcl
@@ -1889,7 +1889,11 @@ proc get_statefile_value {class fd result} {
 
 # check_statefile
 # Check completed/selected state of target/variant $name
-proc check_statefile {class name fd} {
+proc check_statefile {class name {fd {}}} {
+    global target_state_fd
+    if {$fd eq {} && [info exists target_state_fd]} {
+        set fd $target_state_fd
+    }
     seek $fd 0
     while {[gets $fd line] >= 0} {
         if {$line eq "$class: $name"} {
@@ -1901,7 +1905,11 @@ proc check_statefile {class name fd} {
 
 # write_statefile
 # Set target $name completed in the state file
-proc write_statefile {class name fd} {
+proc write_statefile {class name {fd {}}} {
+    global target_state_fd
+    if {$fd eq {} && [info exists target_state_fd]} {
+        set fd $target_state_fd
+    }
     if {[check_statefile $class $name $fd]} {
         return 0
     }
